File: Server/src/Fulfillment/Methods/executeTest.js
const PythonShell = require('python-shell');
const DEV_CONFIG = (process.env.DEVELOPMENT_CONFIG == 'true');
const fLog = '[FULFILLMENT] ';

module.exports.executeTest = function(fileLink, test, test_original, attr, response){
    const options = {
        mode: 'text',
        scriptPath: 'Server/src/Python/',
        args: [`${fileLink}`, `${test}`, `${attr}`]
    };

    PythonShell.run('executeTest.py', options, (err, result) => {
        if (err){
            console.error(`${fLog}ERROR: ${err}`);
        }

        let message;
        switch (test){
            case 'maximum':
            case 'minimum':
            case 'mean':
            case 'std':
            case 'var':{
                message = `The ${test_original} for ${attr} is ${result}`;
            }
                break;
            default: {
                message = `${result}`
            }
        }

        let codeToSend = `import pandas as pd
import urllib


url = ${fileLink}
test_type = ${test}
selected_attr = ${attr}
try:
    data_set = pd.read_csv(url, sep=',', na_values=["?"])
    result = {
        'maximum': data_set[selected_attr].max(),
        'minimum': data_set[selected_attr].min(),
        'mean': data_set[selected_attr].mean(),
        'std': data_set[selected_attr].std(),
        'var': data_set[selected_attr].var()
    }.get(test_type)
    
    if result:
        print(result)
    else:
        print('Something went wrong')
except urllib.error.HTTPError as err:
    if err.code == 404:
    print('ERROR: The provided url is unreachable')
`

        response.send({
            fulfillmentText: message,
            fulfillmentMessages: [
                {
                   platform: 'PLATFORM_UNSPECIFIED',
                   text: {
                       text: [
                           `${result}`
                       ]
                   }
                }
            ],
            payload: {
                code: codeToSend
            }
        });
    });
};

File: Server/src/Fulfillment/Python/executeTest.py
import pandas as pd
import sys
import urllib

if len(sys.argv) > 1:

    if sys.argv[1] != 'null':
        url = sys.argv[1]
    else:
        url = None

    if sys.argv[2] != 'null':
        divider = sys.argv[2]
    else:
        divider = None

    if sys.argv[3] != 'null':
        test_type = sys.argv[3]
    else:
        test_type = None

    if sys.argv[4] != 'null':
        selected_attr = sys.argv[4]
    else:
        selected_attr = None

    try:
        data_set = pd.read_csv(url, sep=divider, na_values=["?"])

        result = {
            'maximum': data_set[selected_attr].max(),
            'minimum': data_set[selected_attr].min(),
            'mean': data_set[selected_attr].mean(),
            'std': data_set[selected_attr].std(),
            'var': data_set[selected_attr].var()
        }.get(test_type)

        print(result)

    except urllib.error.HTTPError as err:
        if err.code == 404:
            print('ERROR: The provided url is unreachable')
else:
    print('Missing parameters')


